src/output-json.c

```diff
@@ -387,12 +391,115 @@ static void JsonAddMetadata(const Packet *p, const Flow *f, json_t *js)
     }
 }
 
+static void JsonAddPayload(json_t *js, const Packet *p, uint16_t flags)
+{
+    if (flags & LOG_JSON_PAYLOAD_BASE64) {
+        unsigned long len = p->payload_len * 2 + 1;
+        uint8_t encoded[len];
+        if (Base64Encode(p->payload, p->payload_len, encoded, &len) == SC_BASE64_OK) {
+            json_object_set_new(js, "payload", json_string((char *)encoded));
+        }
+    }
+
+    if (flags & LOG_JSON_PAYLOAD) {
+        uint8_t printable_buf[p->payload_len + 1];
+        uint32_t offset = 0;
+        PrintStringsToBuffer(printable_buf, &offset,
+                p->payload_len + 1,
+                p->payload, p->payload_len);
+        printable_buf[p->payload_len] = '\0';
+        json_object_set_new(js, "payload_printable", json_string((char *)printable_buf));
+    }
+}
+
+
+
+/* Callback function to pack payload contents from a stream into a buffer
+ * so we can report them in JSON output. */
+static int JsonDumpStreamSegmentCallback(const Packet *p, void *data, const uint8_t *buf, uint32_t buflen)
+{
+    MemBuffer *payload = (MemBuffer *)data;
+    MemBufferWriteRaw(payload, buf, buflen);
+
+    return 1;
+}
+
+
+static void JsonAddStreamDataOneWay(const Packet *p, const Flow *f, json_t *js,
+                                    uint8_t flags, MemBuffer *payload, uint8_t flag)
+{
+    StreamSegmentForEach((const Packet *)p, flag,
+            JsonDumpStreamSegmentCallback,
+            (void *)payload);
+    if (payload->offset) {
+        if (flags & LOG_JSON_PAYLOAD_BASE64) {
+            unsigned long len = payload->offset * 2 + 1;
+            uint8_t encoded[len];
+            Base64Encode(payload->buffer, payload->offset, encoded, &len);
+            json_object_set_new(js, "payload", json_string((char *)encoded));
+        }
+
+        if (flags & LOG_JSON_PAYLOAD) {
+            uint8_t printable_buf[payload->offset + 1];
+            uint32_t offset = 0;
+            PrintStringsToBuffer(printable_buf, &offset,
+                    sizeof(printable_buf),
+                    payload->buffer, payload->offset);
+            json_object_set_new(js, "payload_printable",
+                    json_string((char *)printable_buf));
+        }
+    }
+}
+
+static void JsonAddStreamData(const Packet *p, const Flow *f, json_t *js,
+                              uint8_t flags, MemBuffer *payload)
+{
+    json_t *stream_js = json_object();
+    json_t *sjs = json_object();
+    json_t *cjs = json_object();
+
+    if (!stream_js || !cjs || !sjs)
+        return;
```

Comment:
Shouldn't the non-null items be called with `json_decref`?

Reply:
yes, looks like a good fix.